File: modules/debugger/src/main/java/com/falsepattern/zigbrains/debugger/runner/base/ZigDebugParametersEmitBinaryBase.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.debugger.runner.base;

import com.falsepattern.zigbrains.debugger.Utils;
import com.falsepattern.zigbrains.project.execution.base.ProfileStateBase;
import com.falsepattern.zigbrains.project.toolchain.AbstractZigToolchain;
import com.intellij.execution.ExecutionException;
import com.intellij.openapi.util.io.FileUtil;
import com.jetbrains.cidr.execution.debugger.backend.DebuggerDriverConfiguration;
import lombok.val;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public abstract class ZigDebugParametersEmitBinaryBase<ProfileState extends ProfileStateBase<?>> extends ZigDebugParametersBase<ProfileState> implements PreLaunchAware {
    protected volatile File executableFile;
    private final String kind;
    public ZigDebugParametersEmitBinaryBase(DebuggerDriverConfiguration driverConfiguration, AbstractZigToolchain toolchain, ProfileState profileState, String kind) {
        super(driverConfiguration, toolchain, profileState);
        this.kind = kind;
    }

    private File compileExe()
            throws ExecutionException, Utils.ProcessException {
        final File executableFile;
        val commandLine = profileState.getCommandLine(toolchain, true);
        final Path tmpDir;
        try {
            tmpDir = FileUtil.createTempDirectory("zigbrains_debug", "", true).toPath();
        } catch (IOException e) {
            throw new ExecutionException("Failed to create temporary directory for " + kind + " binary", e);
        }
        val exe = tmpDir.resolve("executable").toFile();
        commandLine.addParameters("-femit-bin=" + exe.getAbsolutePath());
        Utils.executeCommandLineWithErrorChecks(commandLine);
        //Find our binary
        try (val stream = Files.list(tmpDir)){
            executableFile = stream.filter(file -> !file.getFileName().toString().endsWith(".o"))
                                   .map(Path::toFile)
                                   .filter(File::canExecute)
                                   .findFirst()
                                   .orElseThrow(() -> new IOException("No executable file present in temporary directory \"" +
                                                                      tmpDir + "\""));
        } catch (Exception e) {
            throw new ExecutionException("Failed to find compiled binary! " + e.getMessage(), e);
        }
        return executableFile;
    }

    @Override
    public void preLaunch() throws Exception {
        this.executableFile = compileExe();
    }

}

File: cidr/src/main/kotlin/com/falsepattern/zigbrains/debugger/runner/base/ZigDebugParametersEmitBinaryBase.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.debugger.runner.base

import com.falsepattern.zigbrains.debugger.ZigDebugBundle
import com.falsepattern.zigbrains.project.execution.base.ZigProfileState
import com.falsepattern.zigbrains.project.toolchain.AbstractZigToolchain
import com.intellij.execution.ExecutionException
import com.intellij.openapi.util.io.FileUtil
import com.intellij.platform.util.progress.withProgressText
import com.intellij.util.containers.orNull
import com.jetbrains.cidr.execution.debugger.backend.DebuggerDriverConfiguration
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.nio.file.Files
import kotlin.io.path.absolutePathString
import kotlin.io.path.isExecutable

abstract class ZigDebugParametersEmitBinaryBase<ProfileState: ZigProfileState<*>>(
    driverConfiguration: DebuggerDriverConfiguration,
    toolchain: AbstractZigToolchain,
    profileState: ProfileState,
) : ZigDebugParametersBase<ProfileState>(driverConfiguration, toolchain, profileState), PreLaunchAware {
    @Volatile
    protected lateinit var executableFile: File
        private set

    @Throws(ExecutionException::class)
    private suspend fun compileExe(listener: PreLaunchProcessListener): File {
        val commandLine = profileState.getCommandLine(toolchain, true)
        val tmpDir = FileUtil.createTempDirectory("zigbrains_debug", "", true).toPath()

        val exe = tmpDir.resolve("executable")
        commandLine.addParameters("-femit-bin=${exe.absolutePathString()}")

        if (listener.executeCommandLineWithHook(commandLine))
            throw ExecutionException(ZigDebugBundle.message("debug.base.compile.failed.generic"))

        return withContext(Dispatchers.IO) {
            Files.list(tmpDir).use { stream ->
                 stream.filter { !it.fileName.endsWith(".o") }
                     .filter { it.isExecutable() }
                     .findFirst()
                     .map { it.toFile() }
                     .orNull()
            }
        } ?: throw ExecutionException(ZigDebugBundle.message("debug.base.compile.failed.no-exe"))
    }

    @Throws(ExecutionException::class)
    override suspend fun preLaunch(listener: PreLaunchProcessListener) {
        this.executableFile = withProgressText("Compiling executable") {
            compileExe(listener)
        }
    }
}